pandas/core/extensions.py

```diff
@@ -0,0 +1,216 @@
+"""Extend pandas with custom array types.
+"""
+import abc
+import typing as T
+
+import numpy as np
+
+
+class ExtensionDtype(metaclass=abc.ABCMeta):
+    """A custom data type for your array.
+    """
+    @property
+    @abc.abstractmethod
+    def type(self) -> type:
+        """Typically a metaclass inheriting from 'type' with no methods."""
+
+    @property
+    @abc.abstractmethod
+    def base(self) -> np.dtype:
+        # TODO
+        pass
+
+    @property
+    def kind(self) -> str:
+        """A character code (one of 'biufcmMOSUV'), default 'O'
+
+        See Also
+        --------
+        numpy.dtype.kind
+        """
+        return 'O'
+
+    @property
+    @abc.abstractmethod
+    def name(self) -> str:
+        """An string identifying the data type.
+
+        Will be used in, e.g. ``Series.dtype``
+        """
+
+    @property
+    def names(self) -> T.Optional[T.List[str]]:
+        """Ordered list of field names, or None if there are no fields"""
+        return None
+
+    @classmethod
+    def construct_from_string(cls, string: str) -> 'ExtensionDtype':
+        """Attempt to construct this type from a string.
+
+        Parameters
+        ----------
+        string : str
+
+        Returns
+        -------
+        self : instance of 'cls'
+
+        Raises
+        ------
+        TypeError
+
+        Notes
+        -----
+        The default implementation checks if 'string' matches your
+        type's name. If so, it calls your class with no arguments.
+        """
+        if string == cls.name:
+            return cls()
+        else:
+            raise TypeError("Cannot construct a '{}' from "
+                            "'{}'".format(cls, string))
+
+    @classmethod
+    def is_dtype(cls, dtype: T.Union[str, np.dtype]) -> bool:
+        """Check if we match 'dtype'
+
+        Parameters
+        ----------
+        dtype : str or dtype
+
+        Returns
+        -------
+        is_dtype : bool
+
+        Notes
+        -----
+        The default implementation is True if
+
+        1. 'dtype' is a string that returns true for
+           ``cls.construct_from_string``
+        2. 'dtype' is ``cls`` or a subclass of ``cls``.
+        """
+        if isinstance(dtype, str):
+            try:
+                return isinstance(cls.construct_from_string(dtype), cls)
+            except TypeError:
+                return False
+        else:
+            return issubclass(dtype, cls)
+
+
+class ExtensionArray(metaclass=abc.ABCMeta):
```

Comment:
Suppose you want to implement arithmetic or aggregations for your custom dtype (e.g., consider a units dtype). In principle almost any function that does computation might need to be overloaded in a dtype specific way. (This is what NumPy does for ufuncs.)

Would this be done via defining special methods on `ExtensionArray`? Or do we simply say that this isn't possible?

Reply:
At the moment I've ignored arithmetic methods and reductions. I'm OK with saying it isn't possible for now, but am open to allowing it in the future.

Reply:
I've been working on making Series/Index/DataFrame arithmetic methods behave predictably by sharing implementations (in particular datetime and timedelta types).  Eventually the implementations need to get out of `core.ops` and into some type of mixin classes that get mixed into both `Index` subclasses and `Block` subclasses (e.g. `TimestampVector --> DatetimeIndex, DatetimeBlock`).  The base classes for that look a lot like `ExtensionArray` here; it's probably worth figuring out how these relate sooner rather than later.

Reply:
Agreed. If you look at `core/period.py` I had `PeriodArray` inherit from `DatetimeIndexOpsMixin` to get arithmetic ops to work, which doesn't great.

Reply:
>  I had PeriodArray inherit from DatetimeIndexOpsMixin to get arithmetic ops to work, which doesn't feel great.

Which methods did you need to get from DatetimeIndexOpsMixin?  There's a lot of Index-specific stuff in that class that isn't needed for a vector base class.

> But does a Series with periods also rely on those methods? (which for real extensions would be important; and if not, I am not sure we should add it to PeriodArray et al)

@jorisvandenbossche not at the moment, no.  But the idea as I understand it is that `PeriodArray` would get mixed into something like `PeriodBlock`, then Series[Period] would dispatch certain operations to `self._data.blocks[0].__add__` or whatever.